c++/src/kj/async-inl.h

```diff
@@ -166,6 +166,15 @@ class Event {
   // Enqueues this event to happen after all other events have run to completion and there is
   // really nothing left to do except wait for I/O.
 
+  bool isNext();
```

Comment:
I think there's an additional rule around `isNext()` which some of your usages and "TODO"s are missing: It is not safe to rely on `isNext()` unless the code is about to return to the event loop anyway. In particular, it's never safe to rely on `isNext()` while application code is on the stack, because:

1. The application code might manipulate the event loop further before returning to it.
2. If an `if (isNext()) fire();` results in executing some application callback, the application may not be expecting this callback to run synchronously.

Essentially, you can only do `if (isNext()) fire();` inside some other `fire()` callback, or something directly called by some other `fire()` without entering application code.

I believe your optimization for coroutines satisfies this, since if you suspended the coroutine, control would in fact be passed back to the `fire()` function which would then return to the event loop.

For `ChainPromiseNode`, the place you should be applying the optimization is inside its `fire()` method, not inside the constructor nor `onReady()`, both of which are potentially invoked with application code on the stack.

Then again, this means the optimization's usefulness is pretty narrow: If you can only really use it inside `fire()`, but returning from `fire()` is already going to immediately loop to the next event, then how much are you really saving? In the case of coroutines, we're saving the expense of suspending the coroutine, which might be worth it (I have no idea how expensive that is...).

Reply:
Took me a bit to wrap my head around all this, but I agree. I backed out the optimization attempt here: https://github.com/capnproto/capnproto/pull/1104/commits/029e11883fd5e4fce2e795ce9c45be9f3eef0acc